### PR TITLE
Support default values on NamedTuple fields

### LINKED ISSUE #52676: Support default values on NamedTuple fields
## Example

Currently, this code produces the following error message:

```python
import torch
from torch.fx import symbolic_trace
from collections import namedtuple
from typing import Dict, NamedTuple, Optional, Tuple

class Point(NamedTuple):
    x: Optional[torch.Tensor] = None
    y: Optional[torch.Tensor] = None

class M(torch.nn.Module):
    def __init__(self):
        super(M, self).__init__()

    def forward(self, point: Point):
        return point

p = Point(x=torch.rand(3), y=torch.rand(3))
scripted = torch.jit.script(M())
```

```
Traceback (most recent call last):
  File "test/tinytest.py", line 22, in <module>
    scripted = torch.jit.script(M())
  File "/home/ansley/local/miniconda3/envs/pytorch/lib/python3.8/site-packages/torch/jit/_script.py", line 947, in script
    return torch.jit._recursive.create_script_module(
  File "/home/ansley/local/miniconda3/envs/pytorch/lib/python3.8/site-packages/torch/jit/_recursive.py", line 398, in create_script_module
    return create_script_module_impl(nn_module, concrete_type, stubs_fn)
  File "/home/ansley/local/miniconda3/envs/pytorch/lib/python3.8/site-packages/torch/jit/_recursive.py", line 459, in create_script_module_impl
    create_methods_and_properties_from_stubs(concrete_type, method_stubs, property_stubs)
  File "/home/ansley/local/miniconda3/envs/pytorch/lib/python3.8/site-packages/torch/jit/_recursive.py", line 341, in create_methods_and_properties_from_stubs
    concrete_type._create_methods_and_properties(property_defs, property_rcbs, method_defs, method_rcbs, method_defaults)
  File "/home/ansley/local/miniconda3/envs/pytorch/lib/python3.8/site-packages/torch/jit/annotations.py", line 351, in try_ann_to_type
    return torch._C._resolve_type_from_object(ann, loc, fake_rcb)
RuntimeError: 
Default values are currently not supported on NamedTuple fields in TorchScript. Fields with default values: [xy]:
  File "test/tinytest.py", line 17
    def forward(self, point: Point):
                             ~~~~~ <--- HERE
        return point
```

It would be nice if it didn't throw. One of our internal partners ran up against this.

cc @gmagogsfm

### PR DESCRIPTION
**Pull request by ansley**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#54682**

Differential Revision: [D27327241](https://our.internmc.facebook.com/intern/diff/D27327241)

### CODE REVIEW

File: test/test_jit.py [hunk 1]
```diff
@@ -15632,7 +15649,7 @@ def fn(x=None):
         self.checkScript(fn, ((3, 4),))
         self.checkScript(fn, ())
 
-    def test_named_tuple_redefine(self):
+    def test_namedtuple_redefine(self):
```
Comment (gmagogsfm) [inline test/test_jit.py:15652]: Nice job leaving the campground cleaner

File: test/test_jit.py [hunk 2]
```diff
@@ -119,6 +119,9 @@ def canonical(graph):
 def LSTMCellF(input, hx, cx, *params):
     return LSTMCell(input, (hx, cx), *params)
 
+class Point(NamedTuple):
```
Comment (gmagogsfm) [inline test/test_jit.py:122]: I assume you define `Point` here because TorchScript can't compile class that's not defined in global scope? If so, you can actually define this class local to test case and use `make_global` to make TS see it.

File: torch/csrc/jit/python/python_sugared_value.cpp [hunk 3]
```diff
@@ -924,34 +924,22 @@ TypePtr registerNamedTuple(const py::object& obj, const SourceRange& loc) {
   TORCH_INTERNAL_ASSERT(isNamedTupleClass(obj));
   auto qualifiedName = c10::QualifiedName(py::cast<std::string>(
       py::module::import("torch._jit_internal").attr("_qualified_name")(obj)));
-  // Currently don't support default values
```
Comment (gmagogsfm) [inline torch/csrc/jit/python/python_sugared_value.cpp:927]: Love seeing deletions of code like this : )

File: test/test_jit.py [hunk 4]
```diff
@@ -119,6 +119,9 @@ def canonical(graph):
...
+    x: Optional[torch.Tensor] = None
```
Comment (gmagogsfm) [inline test/test_jit.py:123]: Nit: try some default value other than `None` as well for better test coverage

**Review by gmagogsfm — APPROVED** [review]
Looks good, approved with just a few minor comments.

File: test/test_jit.py [hunk 5]
```diff
@@ -14971,6 +15055,57 @@ def foo():
...
+                new_x = x
+            return new_x
+
+        self.checkScript(fn, ((3, 4),))
+        self.checkScript(fn, ())
+
+    def test_namedtuple_redefine(self):
+        global _1, _2
+        _1 = namedtuple('GoogLeNetOutputs', ['logits', 'aux_logits2', 'aux_logits1'])
+        _2 = namedtuple('GoogLeNetOutputs', ['different'])
```
Comment (gmagogsfm) [inline test/test_jit.py:15073]: Is this code copied from serialized model? The `_1` and `_2` seems a bit odd, please name them properly.

File: test/test_jit.py [hunk 6]
```diff
@@ -14971,6 +15055,57 @@ def foo():
...
+
+        with self.assertRaisesRegex(RuntimeError, r'redefine'):
+            @torch.jit.script
+            def foo(x, y):
+                # type: (_1, _2) -> _1
+                return x
```
Comment (gmagogsfm) [inline test/test_jit.py:15079]: Try to minimize amount of logic under `assertRaisesRegex` by defining `foo` outside but scripting it inside.

File: test/test_jit.py [hunk 7]
```diff
@@ -13790,6 +13789,91 @@ def test_error():
...
+        # constructor. Using the factory constructor with TorchScript
+        # TorchScript creates an anonymous `NamedTuple` class instead of
+        # preserving the actual name. For example, the actual generated
+        # signature in this case is:
+        #   graph(%x.1 : NamedTuple(x : Tensor, y : Tensor))
+        # It looks like similar test cases have had this issue as well
+        # (see: `test_namedtuple_python`).
+        FileCheck().check("NamedTuple")   \
+                   .check_next(r"return (%x.1)")    \
+                   .run(fn.graph)
```
Comment (gmagogsfm) [inline test/test_jit.py:13875]: What does the entire graph look like? I am wondering if there are better patterns to check than the word `NamedTuple` and `return`.

File: torch/_jit_internal.py [hunk 8]
```diff
@@ -975,19 +975,22 @@ def _try_get_dispatched_fn(fn):
...
             the_type = torch.jit.annotations.ann_to_type(obj.__annotations__[field], fake_range())
             annotations.append(the_type)
         else:
             annotations.append(torch._C.TensorType.getInferred())
-    return type(obj).__name__, fields, annotations
+    return type(obj).__name__, obj._fields, annotations, defaults
 
 
-def _create_named_tuple(t, unqual_name: str, field_names: List[str]):
+def _create_named_tuple(t, unqual_name: str, field_names: Dict[str, Any]):
```
Comment (gmagogsfm) [inline torch/_jit_internal.py:993]:
```suggestion
def _create_named_tuple(t, unqual_name: str, fields: Dict[str, Any]):
```

Should callsite of this method be updated too? 
https://github.com/pytorch/pytorch/blob/45012da298cdf5e37121080d9f4273fd8a58d070/torch/csrc/jit/python/pybind_utils.h#L727

File: test/test_jit.py (hunk 5 above)
Comment (ansley) [reply test/test_jit.py:15073]: Looks like this was a `rebase` issue. The original code is [here](https://github.com/pytorch/pytorch/blob/master/test/jit/test_typing.py#L535). I'll delete this

File: test/test_jit.py (hunk 7 above)
Comment (ansley) [reply test/test_jit.py:13875]:
```
graph(%x.1 : NamedTuple(x : Tensor, y : Tensor)):
  return (%x.1)
```

Maybe check for the whole first line? I try to avoid using variable names in `FileCheck` in the case that our naming logic changes later, but in this case it might be relevant

File: test/test_jit.py (hunk 6 above)
Comment (ansley) [reply test/test_jit.py:15079]: Deleted (rebase issue, not my code)

File: torch/_jit_internal.py (hunk 8 above)
Comment (ansley) [reply torch/_jit_internal.py:993]: Thanks, I didn't know about this!

File: torch/csrc/jit/serialization/import_source.cpp [hunk 9]
```diff
@@ -638,23 +638,27 @@ struct SourceImporterImpl : public Resolver,
...
       }
-
       const auto assign = Assign(statement);
-      auto name = Var(assign.lhs()).name().name();
-      field_names.emplace_back(std::move(name));
+
+      auto name = Var(Assign(statement).lhs()).name().name();
+
+      fields.emplace_back(
+          std::pair<std::string, IValue>(std::move(name), IValue()));
```
[the remark this answers is not in the document]
Comment (ansley) [reply torch/csrc/jit/serialization/import_source.cpp]: Done, added tests

File: torch/csrc/jit/serialization/import_source.cpp [hunk 10]
```diff
@@ -485,13 +485,13 @@ struct SourceImporterImpl : public Resolver,
...
-          } else if (hook_names.find(def.name().name()) != hook_names.end()) {
-            hook_def_map.emplace(def.name().name(), def);
-            hook_resolver_map.emplace(def.name().name(), shared_from_this());
+          auto def_name = def.name().name();
+          if (pre_hook_names.find(def_name) != pre_hook_names.end()) {
+            pre_hook_def_map.emplace(def_name, def);
+            pre_hook_resolver_map.emplace(def_name, shared_from_this());
+          } else if (hook_names.find(def_name) != hook_names.end()) {
+            hook_def_map.emplace(def_name, def);
+            hook_resolver_map.emplace(def_name, shared_from_this());
```
Comment (gmagogsfm) [inline torch/csrc/jit/serialization/import_source.cpp:494]: This looks like a good clean up, but it is not related to this named tuple functionality in this PR, please make a separate PR for it.

File: torch/csrc/jit/serialization/import_source.cpp [hunk 11]
```diff
@@ -639,23 +639,34 @@ struct SourceImporterImpl : public Resolver,
...
-
       const auto assign = Assign(statement);
-      auto name = Var(assign.lhs()).name().name();
-      field_names.emplace_back(std::move(name));
+
+      auto name = Var(Assign(statement).lhs()).name().name();
+      auto default_val = IValue();
+      if (assign.rhs().present()) {
+        std::vector<IValue> parsed = type_parser.evaluateDefaults(
+            assign.range(), {assign.rhs().get()}, {assign.type().get()});
```
Comment (gmagogsfm) [inline torch/csrc/jit/serialization/import_source.cpp:656]: What's the rationale behind using `assign.range()` instead of `assign.rhs().range()`?

File: torch/csrc/jit/python/python_sugared_value.cpp [hunk 12]
```diff
@@ -930,40 +930,48 @@ TypePtr registerNamedTuple(const py::object& obj, const SourceRange& loc) {
...
+  // have default values for all its fields
+  TORCH_INTERNAL_ASSERT(field_names.size() == field_defaults.size());
+
+  for (size_t i = 0; i < field_names.size(); ++i) {
+    auto ival = IValue();
+    if (!field_defaults[i].is_none()) {
+      auto type = tryToInferType(field_defaults[i]);
+      ival = toIValue(field_defaults[i], type.type());
+      TORCH_CHECK(
+          ival.tagKind() != "Tensor",
```
Comment (gmagogsfm) [inline torch/csrc/jit/python/python_sugared_value.cpp:963]:
```suggestion
          !ival.isTensor(),
```

File: torch/csrc/jit/python/python_sugared_value.cpp [hunk 13]
```diff
@@ -930,40 +930,48 @@ TypePtr registerNamedTuple(const py::object& obj, const SourceRange& loc) {
...
+    fields.emplace_back(std::pair<std::string, IValue>(field_names[i], ival));
+  }
 
-  auto tt = TupleType::createNamed(qualifiedName, fields, annotations);
+  auto tt = TupleType::createNamed(qualifiedName, fields, field_annotations);
   if (auto type = get_python_cu()->get_type(qualifiedName)) {
     TORCH_CHECK(
         type->isSubtypeOf(tt),
-        "Can't to redefine NamedTuple: ",
+        "Can't redefine NamedTuple after creation: ",
```
Comment (gmagogsfm) [inline torch/csrc/jit/python/python_sugared_value.cpp:974]: "after creation" may be a little confusing since it can refer to creation of named tuple instance.
```suggestion
        "Can't redefine NamedTuple: ",
```

File: torch/csrc/jit/python/python_sugared_value.cpp [hunk 14]
```diff
@@ -930,40 +930,48 @@ TypePtr registerNamedTuple(const py::object& obj, const SourceRange& loc) {
...
+
+  for (size_t i = 0; i < field_names.size(); ++i) {
+    auto ival = IValue();
+    if (!field_defaults[i].is_none()) {
+      auto type = tryToInferType(field_defaults[i]);
+      ival = toIValue(field_defaults[i], type.type());
+      TORCH_CHECK(
+          ival.tagKind() != "Tensor",
+          "Tensors are not "
+          "supported as default NamedTuple fields");
```
Comment (gmagogsfm) [inline torch/csrc/jit/python/python_sugared_value.cpp:965]: Maybe state the reason here, we choose not to support it because it is dangerous to use a mutable object as default value.

File: aten/src/ATen/core/type.cpp [hunk 15]
```diff
@@ -735,13 +742,28 @@ TupleTypePtr TupleType::createNamed(
...
+  for (size_t i = 0; i < fields.size(); ++i) {
+    TORCH_CHECK(fields[i].second.tagKind() != "Tensor", "Tensors are "
+                "not supported as default NamedTuple fields");
     arguments.emplace_back(
-        /*name=*/field_names[i],
+        /*name=*/fields[i].first,
         /*type=*/field_types[i],
-        /*N=*/i);
+        /*N=*/i,
+        /*default_value*/fields[i].second);
```
Comment (gmagogsfm) [inline aten/src/ATen/core/type.cpp:766]:
```suggestion
        /*default_value=*/fields[i].second);
```

**Review by gmagogsfm — CHANGES_REQUESTED** [review]

File: test/jit/test_list_dict.py [hunk 16]
```diff
@@ -1771,7 +1771,7 @@ def forward(self, x):
 
 
 class TestNamedTuple(JitTestCase):
-    def test_named_tuple(self):
+    def test_namedtuple(self):
```
[the remark this answers is not in the document]
Comment (ansley) [reply test/jit/test_list_dict.py:1774]: Thanks! We've discussed this on some of the PRs that I've been working on concurrently. In the future, I'll always submit "clean up" PRs separately. This latest round of PRs should be the only ones that still have "clean up" logic in them. :)

File: torch/csrc/jit/frontend/script_type_parser.h [hunk 17]
```diff
@@ -30,7 +30,7 @@ class TORCH_API ScriptTypeParser {
 
   c10::IValue parseClassConstant(const Assign& assign);
 
- private:
+  // private:
```
[the remark this answers is not in the document]
Comment (ansley) [reply torch/csrc/jit/frontend/script_type_parser.h:33]: Yes, thanks

File: torch/csrc/jit/python/pybind_utils.cpp [hunk 18]
```diff
@@ -67,7 +67,7 @@ IValue toIValue(py::handle obj, const TypePtr& type, c10::optional<int32_t> N) {
     case TypeKind::TupleType: {
       py::tuple tuple = py::cast<py::tuple>(obj);
       size_t tuple_size = tuple.size();
-      auto tuple_type = type->cast<TupleType>();
+      TupleTypePtr tuple_type = type->cast<TupleType>();
```
[the remark this answers is not in the document]
Comment (ansley) [reply torch/csrc/jit/python/pybind_utils.cpp:70]: okay

File: torch/csrc/jit/python/pybind_utils.h [hunk 19]
```diff
@@ -713,18 +713,39 @@ inline py::object toPyObject(IValue ivalue) {
...
+      const std::vector<Argument>& tuple_args =
+          tuple->type()->schema()->arguments();
+
+      std::vector<IValue> defaults;
+      auto it = std::find_if(
+          tuple_args.begin(), tuple_args.end(), [](const Argument& arg) {
+            if (arg.default_value()) {
+              return true;
+            }
+            return false;
```
[the remark this answers is not in the document]
Comment (ansley) [reply torch/csrc/jit/python/pybind_utils.h:736]: Much better, thanks

File: torch/_jit_internal.py [hunk 20]
```diff
@@ -984,24 +984,29 @@ def _try_get_dispatched_fn(fn):
...
-    return type(obj).__name__, fields, annotations
+    return type(obj).__name__, obj._fields, annotations, defaults
 
 
-def _create_named_tuple(t, unqual_name: str, field_names: List[str]):
+def _create_named_tuple(t, unqual_name: str, field_names: List[str], defaults: List[Any]):
     # mypy: namedtuple() expects a string literal as the first argument
-    TupleType = collections.namedtuple(unqual_name, field_names)  # type: ignore[misc]
+    if sys.version_info < (3,7,0):
+        TupleType = collections.namedtuple(unqual_name, field_names)  # type: ignore[misc]
```
[the remark this answers is not in the document]
Comment (ansley) [reply torch/_jit_internal.py:1005]: I needed to ignore two separate `mypy` errors that were on the same line. As far as I know, you can only specify "ignore all" (seen here as `type: ignore`) or "ignore one thing" (`type: ignore[COND]`). Please let me know if there's alternate syntax to ignore both, though.

The mypy errors come from: 1) `TupleType` being "redefined" on the next line (`mypy` doesn't recognize the conditional definition, and 2) the `defaults` field not existing in all Python versions (not an issue because of the conditional check)

File: torch/csrc/jit/python/pybind_utils.h [hunk 21]
```diff
@@ -713,18 +713,39 @@ inline py::object toPyObject(IValue ivalue) {
...
+          });
+      std::transform(
```
[the remark this answers is not in the document]
Comment (ansley) [reply torch/csrc/jit/python/pybind_utils.h:738]: Where do you see that `std::transform` is only C++17/20? I thought that the parallelized version of `std::transform` was added in C++17, but `transform` itself has been around for a while. I might be mistaken, though

File: torch/csrc/jit/python/python_sugared_value.cpp (hunk 12 above)
Comment (ansley) [reply torch/csrc/jit/python/python_sugared_value.cpp:963]: Good, thanks!

File: torch/csrc/jit/serialization/import_source.cpp (hunk 10 above)
Comment (ansley) [reply torch/csrc/jit/serialization/import_source.cpp:494]: Sure, reverted

File: torch/csrc/jit/serialization/import_source.cpp (hunk 11 above)
Comment (ansley) [reply torch/csrc/jit/serialization/import_source.cpp:656]: fixed

File: torch/csrc/jit/python/python_sugared_value.cpp (hunk 13 above)
Comment (ansley) [reply torch/csrc/jit/python/python_sugared_value.cpp:974]: Sure, fixed

File: aten/src/ATen/core/type.cpp (hunk 15 above)
Comment (ansley) [reply aten/src/ATen/core/type.cpp:766]: good, thanks

File: torch/_jit_internal.py [hunk 22]
```diff
@@ -984,24 +984,29 @@ def _try_get_dispatched_fn(fn):
...
             annotations.append(torch._C.TensorType.getInferred())
-    return type(obj).__name__, fields, annotations
+    return type(obj).__name__, obj._fields, annotations, defaults
 
 
-def _create_named_tuple(t, unqual_name: str, field_names: List[str]):
+def _create_named_tuple(t, unqual_name: str, field_names: List[str], defaults: List[Any]):
     # mypy: namedtuple() expects a string literal as the first argument
-    TupleType = collections.namedtuple(unqual_name, field_names)  # type: ignore[misc]
+    if sys.version_info < (3,7,0):
```
[the remark this answers is not in the document]
Comment (ansley) [reply torch/_jit_internal.py:1004]: done

File: torch/csrc/jit/python/pybind_utils.h (hunk 21 above)
[the remark this answers is not in the document]
Comment (gmagogsfm) [reply torch/csrc/jit/python/pybind_utils.h:738]: You are right, I misremembered. Just checked cppreference and indeed `std::transform` is available.

File: torch/csrc/jit/python/python_sugared_value.cpp (hunk 13 above)
Comment (gmagogsfm) [reply torch/csrc/jit/python/python_sugared_value.cpp:974]: Doesn't seem changed yet. Maybe because the latest revision is not pushed yet? In general, please push the fix before replying to the comment thread so that the change is visible and reviewer can confirm is the fix is what they meant.

File: torch/csrc/jit/python/python_sugared_value.cpp (hunk 14 above)
Comment (gmagogsfm) [reply torch/csrc/jit/python/python_sugared_value.cpp:965]: Something like this:
```
import torch
from typing import NamedTuple, List, Dict, Optional

t = torch.ones(1)

class Point(NamedTuple):
    x: Optional[List[int]] = None
    y: torch.Tensor = t

print(Point(1))
t[0] = 2
print(Point(1))
```

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27327241).

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27327241).

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27327241).

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27327241).

File: aten/src/ATen/core/type.cpp [hunk 23]
```diff
@@ -742,13 +749,28 @@ TupleTypePtr TupleType::createNamed(
...
+
+TupleTypePtr TupleType::createNamed(
+    const c10::optional<c10::QualifiedName>& qualName,
+    const std::vector<std::pair<std::string, IValue>>& fields,
+    const std::vector<TypePtr>& field_types) {
+  TORCH_INTERNAL_ASSERT(fields.size() == field_types.size());
   std::vector<Argument> arguments;
-  for (size_t i = 0; i < field_names.size(); ++i) {
+  for (size_t i = 0; i < fields.size(); ++i) {
+    TORCH_CHECK(fields[i].second.tagKind() != "Tensor", "Tensors are "
```
Comment (gmagogsfm) [inline aten/src/ATen/core/type.cpp:767]: Consider also state the reason for not supporting Tensor type default value here.

File: torch/_jit_internal.py [hunk 24]
```diff
@@ -993,24 +993,29 @@ def _try_get_dispatched_fn(fn):
...
-    return type(obj).__name__, fields, annotations
+    return type(obj).__name__, obj._fields, annotations, defaults
 
 
-def _create_named_tuple(t, unqual_name: str, field_names: List[str]):
+def _create_named_tuple(t, unqual_name: str, field_names: List[str], defaults: List[Any]):
     # mypy: namedtuple() expects a string literal as the first argument
-    TupleType = collections.namedtuple(unqual_name, field_names)  # type: ignore[misc]
+    if sys.version_info < (3, 7, 0):
+        TupleType = collections.namedtuple(unqual_name, field_names)  # type: ignore[no-redef, misc]
```
Comment (gmagogsfm) [inline torch/_jit_internal.py:1014]: For Python before 3.7, I think there is a hacky way of creating a named tuple with default values:
https://stackoverflow.com/a/18348004

**Review by gmagogsfm — APPROVED** [review]
Looks great, just a couple minor comment.

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27327241).

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27327241).

File: aten/src/ATen/core/type.cpp (hunk 23 above)
Comment (ansley) [reply aten/src/ATen/core/type.cpp:767]: Fixed

File: torch/_jit_internal.py (hunk 24 above)
Comment (ansley) [reply torch/_jit_internal.py:1014]: Thanks!

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27327241).

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27327241).

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27327241).

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27327241).

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27327241).

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27327241).

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27327241).

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27327241).

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27327241).

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27327241).

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27327241).

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27327241).
